File: BinaryTreeLevelOrderTraversalII/bfs.cpp
vector<vector<int>> levelOrderBottom(TreeNode* root) {
    vector<vector<int> > r;
    vector<int> level;
    queue<TreeNode*> q;
    queue<TreeNode*> l;
    if( !root ) return r;
    q.push(root);
    while( !q.empty() ) {
        root = q.front();
        q.pop();
        level.push_back(root->val);

        if( root->left ) 
            l.push(root->left);

        if( root->right )
            l.push(root->right);

        if( q.empty() ) {
            r.push_back(level);
            level.resize(0);
            //copy(l.begin(), l.end(), q.begin());
            while( !l.empty() ) {
                q.push(l.front());
                l.pop();
            }
        }

    }
    reverse(r.begin(),r.end());
    return r;
}
It's very simple. I think we can optimize it by without copying the second queue elements. But I didn't find a way till now
///

vector<vector<int>> levelOrderBottom(TreeNode* root) {
    vector<vector<int>> result;
    if (root == NULL) return result;
    stack<vector<int>> level_tmp;
    queue<TreeNode*> q;
    q.push(root);
    while (!q.empty()) {
        vector<int> level;
        size_t size = q.size();
        for (int i = 0; i < size; ++i) {
            TreeNode* node = q.front(); q.pop();
            level.push_back(node->val);
            if (node->left != NULL) q.push(node->left);
            if (node->right != NULL) q.push(node->right);
        }
        level_tmp.emplace(move(level));
    }
    while (!level_tmp.empty()) {
        result.emplace_back(move(level_tmp.top()));
        level_tmp.pop();
    }
    return vector<vector<int>>(result);
}


///anther
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        vector<vector<int>> ret;
        if (root == NULL) return ret;
        queue<TreeNode *> q;
        q.push(root);
        while (!q.empty()) {
            int l = q.size();
            vector<int> curlayer;
            for (int i = 0; i < l; i++) {
                TreeNode *cur = q.front();
                q.pop();
                curlayer.push_back(cur->val);
                if (cur->left != NULL) q.push(cur->left);
                if (cur->right != NULL) q.push(cur->right);
            }
            ret.push_back(curlayer);
        }
        reverse(ret.begin(), ret.end());
        return ret;
    }
};

File: BinaryTreeLevelOrderTraversalII/bfs.py
Use a List as Queue, then do the BFS iteratively.

# @param root, a tree node
# @return a list of lists of integers
def levelOrderBottom(self, root):
    if root == None:
        return []
    ans = []
    cur = []
    q = [root, None]
    while len(q) > 0:
        node = q.pop(0)
        if node == None:
            if len(cur) > 0:
                ans.insert(0, cur)
                cur = []
                q.append(None)
        else:
            cur.append(node.val)
            if node.left != None:
                q.append(node.left)
            if node.right != None:
                q.append(node.right)
    return ans
Donot use Queue, use two Lists to keep the current level and next level, the do the BFS iteratively.

# @param root, a tree node
# @return a list of lists of integers
def levelOrderBottom(self, root):
    if root == None:
        return []
    ans = []
    cur = [root]
    while len(cur) > 0:
        vals = []
        next = []
        for x in cur:
            vals.append(x.val)
            if x.left != None:
                next.append(x.left)
            if x.right != None:
                next.append(x.right)
        ans.insert(0, vals)
        cur = next
    return ans
Donot use Queue, use two Lists to keep the current level and next level, the do the BFS recursively.

# @param root, a tree node
# @return a list of lists of integers
def levelOrderBottom(self, root):
    if root == None:
        return []
    ans = []
    cur = [root]
    self.traverse(ans, cur)
    return ans


def traverse(self, ans, cur):
    if len(cur) == 0:
        return

    next = []
    for x in cur:
        if x.left != None:
            next.append(x.left)
        if x.right != None:
            next.append(x.right)

    self.traverse(ans, next)

    curVal = []
    for x in cur:
        curVal.append(x.val)
    ans.append(curVal)
